[ews-build] Access the full 'results' endpoint and compare actual and given results

briannafan opened the pull request:
#### 4b1c5f064280c803cb0cc02ca448571fb973b923
<pre>
[ews-build] Access the full &apos;results&apos; endpoint and compare actual and given results
<a href="https://bugs.webkit.org/show_bug.cgi?id=282659">https://bugs.webkit.org/show_bug.cgi?id=282659</a>
<a href="https://rdar.apple.com/139321333">rdar://139321333</a>

Reviewed by Jonathan Bedard.

Add support for safer-cpp-checks suite.
Add access to the &apos;api/results&apos; endpoint for suites and individual tests.
Add a comparison method for actual and user-given results per test.

* Tools/CISupport/ews-build/results_db.py:
(ResultsDatabase): Added safer-cpp-checks and fixed typo.
(ResultsDatabase.make_request): Added. Specify which endpoint to request.
(ResultsDatabase.get_results): Added. Has the option to specify a test to
    retrieve results for. Otherwise, return results for the suite.
(ResultsDatabase.is_test_pre_existing_failure): Fixed typo.
(ResultsDatabase.does_result_match): Added. Returns whether a test result is
    equal to a given result. User can specify a default value for a test if
    results do not exist.

Canonical link: <a href="https://commits.webkit.org/286573@main">https://commits.webkit.org/286573@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ab36aeaca20d96bbda7d2fc6c2229a68ef847ab

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76239 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55269 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29141 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80745 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27511 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78355 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64413 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3565 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59779 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17916 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79306 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49682 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65478 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40124 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47079 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22964 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25833 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68209 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23299 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82204 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3611 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2358 "Found 1 new test failure: imported/w3c/web-platform-tests/fetch/stale-while-revalidate/fetch.any.serviceworker.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68004 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/75916 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3765 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65450 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67315 "Found 1 new API test failure: /TestWebKit:WebKit.GetInjectedBundleInitializationUserDataCallback (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11278 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9387 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11820 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3559 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3582 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7011 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5340 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/CISupport/ews-build/results_db.py:
```diff
@@ -88,13 +88,78 @@ def get_results_summary(cls, test, commit=None, configuration=None, logger=None,
+
+    @classmethod
+    @defer.inlineCallbacks
+    def get_results(cls, test, commit=None, configuration=None, logger=None, suite=None, default=None):
```
I think `get_results`, `get_suite_results ` and `get_results_summary` can share much more code. They're all basically identical save the URL endpoint they're hitting, right?

briannafan replied:
Good point! I combined `get_results` and `get_suite_results` into one function and made `test` an optional argument. Now `get_results` and `get_results_summary` call `make_request` by passing in their arguments plus which endpoint to use.

JonWBedard commented on Tools/CISupport/ews-build/results_db.py:
```diff
@@ -103,6 +168,28 @@ def is_test_pre_existing_failure(cls, test, commit=None, configuration=None, sui
 
+    @classmethod
+    @defer.inlineCallbacks
+    def does_result_match(cls, test, result_type=None, commit=None, configuration=None, suite=None, default=None):
```
Where is this actually used?

briannafan replied:
This will be used in `FindUnexpectedSaferCPPResults` (it's where I tested this function in the build I linked). I'll specify the exact spot once I update the PR for those changes.

briannafan committed:
[ews-build] Access the full 'results' endpoint and compare actual and given results
https://bugs.webkit.org/show_bug.cgi?id=282659
rdar://139321333

Reviewed by Jonathan Bedard.

Add support for safer-cpp-checks suite.
Add access to the 'api/results' endpoint for suites and individual tests.
Add a comparison method for actual and user-given results per test.

* Tools/CISupport/ews-build/results_db.py:
(ResultsDatabase): Added safer-cpp-checks and fixed typo.
(ResultsDatabase.make_request): Added. Specify which endpoint to request.
(ResultsDatabase.get_results): Added. Has the option to specify a test to
    retrieve results for. Otherwise, return results for the suite.
(ResultsDatabase.is_test_pre_existing_failure): Fixed typo.
(ResultsDatabase.does_result_match): Added. Returns whether a test result is
    equal to a given result. User can specify a default value for a test if
    results do not exist.

Canonical link: https://commits.webkit.org/286573@main
